protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+And we can literally "chain", by commiting to the pre-state and new input, to produce the post-state.
+
+##### How do we chain the steps?
+
+We define the output as the hash of the pre-state and the added input-configurtion.
+This makes each step traceable and deterministic.
+
+Note that while steps are "chained" together,
+we can fork the deployment at any stage: share common steps, then expand into different settings.
+So it is not a single deployment sequence, but more like a tree of possible deployments.
+
+##### How do we keep the cache reasonably sized?
+
+Produced increments might take up more space than desired:
+for simplicity, dumping the whole state-allocs each step would be nice.
+
+But if there are many configuration tweaks, and many steps, it may produce too many of these files.
+
+The devnet L1 allocs file today is ~650 KB.
+The devnet L2 allocs file today is 9.1 MB: the majority of this data are L2-predeploy proxy placeholders.
+
+To reduce the size we could make the output incremental also:
+an output can be JSON, and each account entry overwrites the contents of the previous account entry.
+For simplicity we avoid storage-merging, and just overwrite accounts as a whole.
+We do not have contracts with particularly many storage entries, nor do the medium-sized ones change with every step. 
+
+By performing merges this way, we can also make hardforks clean:
+e.g. we can produce a state for fork A, and then fork it and apply the upgrade to fork B, producing a tiny diff.
+
+The smaller the files, the more we can afford, and the higher the limit
+for configuration test-coverage is.
+
+##### How do we not miss the cache with every redeployment?
+
+It is important to omit the "genesis timestamp" from the cache-key, and use some fixed value as input instead:
+every real-time test will need a different timestamp, so embedding it as input would break all caching.
+Luckily, only the legacy L2-output-oracle L1 contract enshrines anything,
+and the L2 genesis block is easily modified before chain-specs are generated.
+
+##### How does this work with production deployments?
+
+The caching, chaining, and commiting is not a part of production.
+As it is against a live chain, not a local state-dump.
+
+However, the isolated configuration and deployment-sub-routines do enable
+a deployer to plug into the superchain contracts, while customizing just what they need,
+without having to untangle a monolithic single deployment script.
+
+Ideally this is performed through the OPStackManager.
+
+### Result
+
+This approach enables us to get rid of "mega-feature" env-vars (Plasma, fault-proofs, Interop)
+as well as allocs hardfork-suffixes: everything is configuration,
+and we can just fork of existing configuration to tweak something.
+
+The main changes to the deploy/genesis-scripting are to:
+1. Get rid of L1 and L2 "allocs": every incremental output is a small "allocs", identified by the hash of inputs.
```

Comment:
Does this mean we're changing the genesis file format used by the node, or just changing the development format, and it gets postprocessed into the standard format for production deploys?

Reply:
Changing an intermediate development format only. The solidity scripts would output a regular state-dump, the Go op-chain-ops tooling would turn that into cached state diffs, and the genesis-generation can then load those diffs into an actual full genesis.